File: app/src/main/java/com/ireddragonicy/konabessnext/ui/adapters/DtsTreeAdapter.java
package com.ireddragonicy.konabessnext.ui.adapters;

import android.graphics.Color;
import android.text.Editable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextWatcher;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ireddragonicy.konabessnext.R;
import com.ireddragonicy.konabessnext.model.dts.DtsNode;
import com.ireddragonicy.konabessnext.model.dts.DtsProperty;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class DtsTreeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements StickyHeaderInterface {

    private static final int TYPE_NODE = 0;
    private static final int TYPE_PROPERTY = 1;

    private final List<TreeItem> visibleItems = new ArrayList<>();
    private DtsNode rootNode;
    private RecyclerView recyclerView;

    // Search State
    private String searchQuery = "";
    private final List<TreeItem> searchResults = new ArrayList<>();
    private int currentSearchIndex = -1;

    public DtsTreeAdapter() {
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        this.recyclerView = recyclerView;
    }

    public void setRootNode(DtsNode root) {
        this.rootNode = root;
        rebuildVisibleList();
    }

    private void rebuildVisibleList() {
        visibleItems.clear();
        if (rootNode != null) {
            // Use iterative approach to prevent StackOverflow on deep trees
            Stack<NodeDepth> stack = new Stack<>();
            // Push root in reverse order effectively? No, standard preorder.
            // But we want to iterate properties then children.
            // To maintain order in a stack: push LastChild ... FirstChild, then Properties.

            // Actually, we can just linearize normally.
            // But doing it iteratively for "visible items" which depends on isExpanded.

            // Let's use a custom iterative function that mimics the recursive one.
            Stack<NodeDepth> traverseStack = new Stack<>();
            traverseStack.push(new NodeDepth(rootNode, 0));

            // The issue with strict stack processing is it does depth-first.
            // But we need to process "Node itself" -> "Properties" -> "Children".
            // If we push Node, pop it, add to list.
            // If expanded: push Children (reverse order), push Properties (reverse order).

            while (!traverseStack.isEmpty()) {
                NodeDepth wrapper = traverseStack.pop();
                DtsNode node = wrapper.node;
                int depth = wrapper.depth;

                visibleItems.add(new TreeItem(node, depth));

                if (node.isExpanded) {
                    // Push children in reverse so they come out in order
                    for (int i = node.children.size() - 1; i >= 0; i--) {
                        traverseStack.push(new NodeDepth(node.children.get(i), depth + 1));
                    }

                    // Push properties in reverse
                    for (int i = node.properties.size() - 1; i >= 0; i--) {
                        DtsProperty prop = node.properties.get(i);
                        // We can't push Property to NodeDepth stack easily if NodeDepth only takes
                        // DtsNode.
                        // We need a generic wrapper or handle properties differently.
                        // Let's handle properties directly here:
                        // But wait, the list order is Node -> Properties -> Children.
                        // If we process children later, we need to add properties NOW.
                        // So we don't need to stack properties recursively (they are leaf items in this
                        // view model - no children).
                        // BUT we need to add them AFTER the node, BEFORE the children.
                        // And since we pop children later, we should add properties to visibleItems
                        // IMMEDIATELY.
                    }

                    // Correct iterative logic:
                    // 1. Add current Node to visibleItems.
                    // 2. Add all its Properties to visibleItems immediately.
                    // 3. Push Children to stack (reverse order).

                    for (DtsProperty prop : node.properties) {
                        visibleItems.add(new TreeItem(prop, depth + 1));
                    }
                }
            }
        }
        notifyDataSetChanged();
    }

    // Helper for stack
    private static class NodeDepth {
        DtsNode node;
        int depth;

        NodeDepth(DtsNode n, int d) {
            this.node = n;
            this.depth = d;
        }
    }

    // ============================================================================================
    // Search Logic
    // ============================================================================================

    public boolean search(String query) {
        this.searchQuery = query == null ? "" : query;
        this.searchResults.clear();
        this.currentSearchIndex = -1;

        if (this.searchQuery.isEmpty()) {
            for (TreeItem item : visibleItems) {
                item.isMatch = false;
            }
            notifyDataSetChanged();
            return false;
        }

        if (rootNode == null)
            return false;

        Stack<DtsNode> path = new Stack<>();
        expandForMatches(rootNode, query.toLowerCase(), path);

        rebuildVisibleList();

        for (TreeItem item : visibleItems) {
            boolean isMatch = false;
            if (item.item instanceof DtsNode) {
                if (((DtsNode) item.item).name.toLowerCase().contains(query.toLowerCase())) {
                    isMatch = true;
                }
            } else if (item.item instanceof DtsProperty) {
                DtsProperty prop = (DtsProperty) item.item;
                if (prop.name.toLowerCase().contains(query.toLowerCase()) ||
                        prop.getDisplayValue().toLowerCase().contains(query.toLowerCase())) {
                    isMatch = true;
                }
            }

            item.isMatch = isMatch;
            if (isMatch) {
                searchResults.add(item);
            }
        }

        if (!searchResults.isEmpty()) {
            currentSearchIndex = 0;
            scrollToResult(searchResults.get(0));
            notifyDataSetChanged();
            return true;
        } else {
            notifyDataSetChanged();
            return false;
        }
    }

    private boolean expandForMatches(DtsNode node, String query, Stack<DtsNode> path) {
        // This is still recursive. For huge trees, we should make this iterative too.
        // But for now, let's rely on stack depth being larger than typical tree depth
        // (usually < 50 for DTS).
        // The display recursion was the main culprit if expanding ALL nodes.

        // Actually, if we had a parsing error creating infinite loop, even iterative
        // would hang (infinite loop) vs crash (stack overflow).
        // But at least it won't crash immediately.

        path.push(node);
        boolean hasMatch = false;

        if (node.name.toLowerCase().contains(query))
            hasMatch = true;

        for (DtsProperty prop : node.properties) {
            if (prop.name.toLowerCase().contains(query) || prop.getDisplayValue().toLowerCase().contains(query)) {
                hasMatch = true;
            }
        }

        for (DtsNode child : node.children) {
            if (expandForMatches(child, query, path)) {
                hasMatch = true;
            }
        }

        if (hasMatch) {
            node.isExpanded = true;
        }

        path.pop();
        return hasMatch;
    }

    public boolean nextSearchResult() {
        if (searchResults.isEmpty())
            return false;
        currentSearchIndex = (currentSearchIndex + 1) % searchResults.size();
        scrollToResult(searchResults.get(currentSearchIndex));
        notifyDataSetChanged();
        return true;
    }

    public boolean previousSearchResult() {
        if (searchResults.isEmpty())
            return false;
        currentSearchIndex = (currentSearchIndex - 1 + searchResults.size()) % searchResults.size();
        scrollToResult(searchResults.get(currentSearchIndex));
        notifyDataSetChanged();
        return true;
    }

    public void clearSearch() {
        search(null);
    }

    private void scrollToResult(TreeItem item) {
        int index = visibleItems.indexOf(item);
        if (index != -1 && recyclerView != null) {
            recyclerView.scrollToPosition(index);
            ((LinearLayoutManager) recyclerView.getLayoutManager()).scrollToPositionWithOffset(index, 200);
        }
    }

    // ============================================================================================

    @Override
    public int getItemViewType(int position) {
        return visibleItems.get(position).item instanceof DtsNode ? TYPE_NODE : TYPE_PROPERTY;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        if (viewType == TYPE_NODE) {
            View v = inflater.inflate(R.layout.item_dts_visual_node, parent, false);
            return new NodeViewHolder(v);
        } else {
            View v = inflater.inflate(R.layout.item_dts_visual_property, parent, false);
            return new PropertyViewHolder(v);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        TreeItem item = visibleItems.get(position);
        if (holder instanceof NodeViewHolder) {
            ((NodeViewHolder) holder).bind((DtsNode) item.item, item.depth, item.isMatch);
        } else if (holder instanceof PropertyViewHolder) {
            ((PropertyViewHolder) holder).bind((DtsProperty) item.item, item.depth, item.isMatch);
        }
    }

    @Override
    public int getItemCount() {
        return visibleItems.size();
    }

    // ============================================================================================
    // Sticky Header Implementation
    // ============================================================================================

    @Override
    public int getHeaderPositionForItem(int itemPosition) {
        if (itemPosition >= visibleItems.size())
            return -1;

        // If current item is a Node, it is its own header (start of section)
        if (getItemViewType(itemPosition) == TYPE_NODE) {
            return itemPosition;
        }

        // Otherwise, search backwards for the first Node (parent)
        for (int i = itemPosition - 1; i >= 0; i--) {
            if (getItemViewType(i) == TYPE_NODE) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public int getHeaderLayout(int headerPosition) {
        return R.layout.item_dts_visual_node;
    }

    @Override
    public void bindHeaderData(View headerView, int headerPosition) {
        if (headerPosition != -1 && headerPosition < visibleItems.size()) {
            TreeItem item = visibleItems.get(headerPosition);
            if (item.item instanceof DtsNode) {
                // Reuse NodeViewHolder logic but for a detached view
                // We create a temporary holder just for binding
                NodeViewHolder holder = new NodeViewHolder(headerView);
                // Bind with current state.
                // Note: Click listeners on sticky header usually don't work with simple
                // ItemDecoration drawing.
                // Ideally user clicks the "real" item if it's visible, or we need touch
                // interception.
                // For "Sticky Note" visual context, just binding data is enough.
                holder.bind((DtsNode) item.item, item.depth, item.isMatch);

                // Override text to show full path
                holder.nameText.setText(((DtsNode) item.item).getFullPath());

                // Force opaque background for sticky header so text doesn't bleed through
                // Use default window background or surface color
                headerView.setBackgroundColor(Color.parseColor("#FF202020")); // Dark grey/black for dark mode
                // Ideally resolving ?attr/colorSurface would be better but requires Context

                // Try resolving colorSurface
                android.util.TypedValue outValue = new android.util.TypedValue();
                headerView.getContext().getTheme().resolveAttribute(com.google.android.material.R.attr.colorSurface,
                        outValue, true);
                if (outValue.resourceId != 0) {
                    headerView.setBackgroundResource(outValue.resourceId);
                } else {
                    headerView.setBackgroundColor(outValue.data); // If it's a color value
                }

                // Ensure it's opaque if surface is transparent (rare)
                // Also, let's add a small elevation/shadow line?
                headerView.setElevation(8f);
            }
        }
    }

    @Override
    public boolean isHeader(int itemPosition) {
        return getItemViewType(itemPosition) == TYPE_NODE;
    }

    @Override
    public void onHeaderClicked(int headerPosition) {
        if (headerPosition != -1 && headerPosition < visibleItems.size()) {
            TreeItem item = visibleItems.get(headerPosition);
            if (item.item instanceof DtsNode) {
                DtsNode node = (DtsNode) item.item;
                node.isExpanded = !node.isExpanded;
                rebuildVisibleList();

                // If we collapsed it, we might be way down the list. Scroll to the header.
                if (!node.isExpanded) {
                    scrollToResult(item);
                }
            }
        }
    }

    // ============================================================================================
    // View Holders
    // ============================================================================================

    class NodeViewHolder extends RecyclerView.ViewHolder {
        private final View indentation;
        private final ImageView expandIcon;
        private final TextView nameText;

        public NodeViewHolder(@NonNull View itemView) {
            super(itemView);
            indentation = itemView.findViewById(R.id.node_indentation);
            expandIcon = itemView.findViewById(R.id.btn_expand);
            nameText = itemView.findViewById(R.id.node_name);

            itemView.setOnClickListener(v -> toggleExpand());
            expandIcon.setOnClickListener(v -> toggleExpand());
        }

        private void toggleExpand() {
            int pos = getAdapterPosition();
            if (pos == RecyclerView.NO_POSITION)
                return;

            TreeItem item = visibleItems.get(pos);
            DtsNode node = (DtsNode) item.item;
            node.isExpanded = !node.isExpanded;

            rebuildVisibleList();
        }

        public void bind(DtsNode node, int depth, boolean isMatch) {
            indentation.getLayoutParams().width = depth * 40;
            indentation.requestLayout();

            if (isMatch && !searchQuery.isEmpty()) {
                nameText.setText(highlightText(node.name, searchQuery));
            } else {
                nameText.setText(node.name);
            }
            expandIcon.setRotation(node.isExpanded ? 0 : -90);

            if (currentSearchIndex != -1 && searchResults.size() > currentSearchIndex &&
                    searchResults.get(currentSearchIndex).item == node) {
                itemView.setBackgroundColor(Color.parseColor("#4Dffeb3b"));
            } else {
                android.util.TypedValue outValue = new android.util.TypedValue();
                itemView.getContext().getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue,
                        true);
                itemView.setBackgroundResource(outValue.resourceId);
            }
        }
    }

    class PropertyViewHolder extends RecyclerView.ViewHolder {
        private final View indentation;
        private final TextView keyText;
        private final EditText valueInput;
        private TextWatcher currentWatcher;

        public PropertyViewHolder(@NonNull View itemView) {
            super(itemView);
            indentation = itemView.findViewById(R.id.prop_indentation);
            keyText = itemView.findViewById(R.id.prop_key);
            valueInput = itemView.findViewById(R.id.prop_value);
        }

        public void bind(DtsProperty prop, int depth, boolean isMatch) {
            indentation.getLayoutParams().width = depth * 40;
            indentation.requestLayout();

            if (isMatch && !searchQuery.isEmpty()) {
                keyText.setText(highlightText(prop.name, searchQuery));
            } else {
                keyText.setText(prop.name);
            }

            if (currentWatcher != null) {
                valueInput.removeTextChangedListener(currentWatcher);
            }

            String displayVal = prop.getDisplayValue();
            valueInput.setText(displayVal);

            currentWatcher = new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                }

                @Override
                public void afterTextChanged(Editable s) {
                    prop.updateFromDisplayValue(s.toString());
                }
            };
            valueInput.addTextChangedListener(currentWatcher);

            if (currentSearchIndex != -1 && searchResults.size() > currentSearchIndex &&
                    searchResults.get(currentSearchIndex).item == prop) {
                itemView.setBackgroundColor(Color.parseColor("#4Dffeb3b"));
            } else {
                android.util.TypedValue outValue = new android.util.TypedValue();
                itemView.getContext().getTheme().resolveAttribute(android.R.attr.selectableItemBackground, outValue,
                        true);
                itemView.setBackgroundResource(outValue.resourceId);
            }
        }
    }

    private SpannableString highlightText(String text, String query) {
        SpannableString spannable = new SpannableString(text);
        if (query == null || query.isEmpty())
            return spannable;

        String lowerText = text.toLowerCase();
        String lowerQuery = query.toLowerCase();

        int start = lowerText.indexOf(lowerQuery);
        while (start != -1) {
            int end = start + lowerQuery.length();
            spannable.setSpan(new BackgroundColorSpan(Color.parseColor("#ffeb3b")), start, end,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            spannable.setSpan(new ForegroundColorSpan(Color.BLACK), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            start = lowerText.indexOf(lowerQuery, end);
        }
        return spannable;
    }

    static class TreeItem {
        Object item; // DtsNode or DtsProperty
        int depth;
        boolean isMatch = false;

        public TreeItem(Object item, int depth) {
            this.item = item;
            this.depth = depth;
        }
    }
}

File: app/src/main/java/com/ireddragonicy/konabessnext/ui/adapters/DtsTreeAdapter.kt
package com.ireddragonicy.konabessnext.ui.adapters

import android.graphics.Color
import android.text.Editable
import android.text.Spannable
import android.text.SpannableString
import android.text.TextWatcher
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ireddragonicy.konabessnext.R
import com.ireddragonicy.konabessnext.model.dts.DtsNode
import com.ireddragonicy.konabessnext.model.dts.DtsProperty
import java.util.ArrayList
import java.util.Stack

class DtsTreeAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>(), StickyHeaderInterface {

    private val visibleItems = ArrayList<TreeItem>()
    private var rootNode: DtsNode? = null
    private var recyclerView: RecyclerView? = null

    // Search State
    private var searchQuery = ""
    private val searchResults = ArrayList<TreeItem>()
    private var currentSearchIndex = -1

    companion object {
        private const val TYPE_NODE = 0
        private const val TYPE_PROPERTY = 1
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    fun setRootNode(root: DtsNode) {
        this.rootNode = root
        rebuildVisibleList()
    }

    private fun rebuildVisibleList() {
        visibleItems.clear()
        if (rootNode != null) {
            // Use iterative approach to prevent StackOverflow on deep trees
            val traverseStack = Stack<NodeDepth>()
            traverseStack.push(NodeDepth(rootNode!!, 0))

            while (!traverseStack.isEmpty()) {
                val wrapper = traverseStack.pop()
                val node = wrapper.node
                val depth = wrapper.depth

                visibleItems.add(TreeItem(node, depth))

                if (node.isExpanded) {
                    // Push children in reverse so they come out in order
                    for (i in node.children.indices.reversed()) {
                        traverseStack.push(NodeDepth(node.children[i], depth + 1))
                    }

                    // Add properties immediately (reverse order not needed if adding directly, but standard loop)
                    // Wait, original logic: added properties THEN pushed children.
                    // Stack processing Order: LIFO.
                    // We popped Node. Added Node.
                    // If we push Child 2, Child 1.
                    // Stack: [Child 2, Child 1]. Pop Child 1. Process.
                    // Effectively processing children in order.
                    // Properties should be processed BEFORE children?
                    // Original code:
                    // visibleItems.add(node)
                    // Pushed children.
                    // Added properties to visibleItems.
                    // Then loop continues (processes next stack item = first child).
                    // So visible items order: Node, Properties, Child 1, Child 1 props...
                    // Correct.

                    for (prop in node.properties) {
                        visibleItems.add(TreeItem(prop, depth + 1))
                    }
                }
            }
        }
        notifyDataSetChanged()
    }

    // Helper for stack
    private class NodeDepth(val node: DtsNode, val depth: Int)

    // ============================================================================================
    // Search Logic
    // ============================================================================================

    fun search(query: String?): Boolean {
        this.searchQuery = query ?: ""
        this.searchResults.clear()
        this.currentSearchIndex = -1

        if (this.searchQuery.isEmpty()) {
            for (item in visibleItems) {
                item.isMatch = false
            }
            notifyDataSetChanged()
            return false
        }

        if (rootNode == null) return false

        val path = Stack<DtsNode>()
        expandForMatches(rootNode!!, searchQuery.lowercase(java.util.Locale.ROOT), path)

        rebuildVisibleList()

        for (item in visibleItems) {
            var isMatch = false
            if (item.item is DtsNode) {
                if ((item.item as DtsNode).name.lowercase(java.util.Locale.ROOT).contains(searchQuery.lowercase(java.util.Locale.ROOT))) {
                    isMatch = true
                }
            } else if (item.item is DtsProperty) {
                val prop = item.item as DtsProperty
                if (prop.name.lowercase(java.util.Locale.ROOT).contains(searchQuery.lowercase(java.util.Locale.ROOT)) ||
                    prop.displayValue.lowercase(java.util.Locale.ROOT).contains(searchQuery.lowercase(java.util.Locale.ROOT))
                ) {
                    isMatch = true
                }
            }

            item.isMatch = isMatch
            if (isMatch) {
                searchResults.add(item)
            }
        }

        return if (searchResults.isNotEmpty()) {
            currentSearchIndex = 0
            scrollToResult(searchResults[0])
            notifyDataSetChanged()
            true
        } else {
            notifyDataSetChanged()
            false
        }
    }

    private fun expandForMatches(node: DtsNode, query: String, path: Stack<DtsNode>): Boolean {
        path.push(node)
        var hasMatch = false

        if (node.name.lowercase(java.util.Locale.ROOT).contains(query)) hasMatch = true

        for (prop in node.properties) {
            if (prop.name.lowercase(java.util.Locale.ROOT).contains(query) || prop.displayValue.lowercase(java.util.Locale.ROOT).contains(query)) {
                hasMatch = true
            }
        }

        for (child in node.children) {
            if (expandForMatches(child, query, path)) {
                hasMatch = true
            }
        }

        if (hasMatch) {
            node.isExpanded = true
        }

        path.pop()
        return hasMatch
    }

    fun nextSearchResult(): Boolean {
        if (searchResults.isEmpty()) return false
        currentSearchIndex = (currentSearchIndex + 1) % searchResults.size
        scrollToResult(searchResults[currentSearchIndex])
        notifyDataSetChanged()
        return true
    }

    fun previousSearchResult(): Boolean {
        if (searchResults.isEmpty()) return false
        currentSearchIndex = (currentSearchIndex - 1 + searchResults.size) % searchResults.size
        scrollToResult(searchResults[currentSearchIndex])
        notifyDataSetChanged()
        return true
    }

    fun clearSearch() {
        search(null)
    }

    private fun scrollToResult(item: TreeItem) {
        val index = visibleItems.indexOf(item)
        if (index != -1 && recyclerView != null) {
            recyclerView!!.scrollToPosition(index)
            (recyclerView!!.layoutManager as LinearLayoutManager).scrollToPositionWithOffset(index, 200)
        }
    }

    // ============================================================================================

    override fun getItemViewType(position: Int): Int {
        return if (visibleItems[position].item is DtsNode) TYPE_NODE else TYPE_PROPERTY
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return if (viewType == TYPE_NODE) {
            val v = inflater.inflate(R.layout.item_dts_visual_node, parent, false)
            NodeViewHolder(v)
        } else {
            val v = inflater.inflate(R.layout.item_dts_visual_property, parent, false)
            PropertyViewHolder(v)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = visibleItems[position]
        if (holder is NodeViewHolder) {
            holder.bind(item.item as DtsNode, item.depth, item.isMatch)
        } else if (holder is PropertyViewHolder) {
            holder.bind(item.item as DtsProperty, item.depth, item.isMatch)
        }
    }

    override fun getItemCount(): Int {
        return visibleItems.size
    }

    // ============================================================================================
    // Sticky Header Implementation
    // ============================================================================================

    override fun getHeaderPositionForItem(itemPosition: Int): Int {
        if (itemPosition >= visibleItems.size) return -1

        // If current item is a Node, it is its own header (start of section)
        if (getItemViewType(itemPosition) == TYPE_NODE) {
            return itemPosition
        }

        // Otherwise, search backwards for the first Node (parent)
        for (i in itemPosition - 1 downTo 0) {
            if (getItemViewType(i) == TYPE_NODE) {
                return i
            }
        }
        return -1
    }

    override fun getHeaderLayout(headerPosition: Int): Int {
        return R.layout.item_dts_visual_node
    }

    override fun bindHeaderData(headerView: View, headerPosition: Int) {
        if (headerPosition != -1 && headerPosition < visibleItems.size) {
            val item = visibleItems[headerPosition]
            if (item.item is DtsNode) {
                val node = item.item as DtsNode
                // Reuse NodeViewHolder logic but for a detached view
                // We create a temporary holder just for binding
                val holder = NodeViewHolder(headerView)
                holder.bind(node, item.depth, item.isMatch)

                // Override text to show full path
                holder.nameText.text = node.getFullPath()

                // Force opaque background for sticky header
                val outValue = TypedValue()
                headerView.context.theme.resolveAttribute(com.google.android.material.R.attr.colorSurface, outValue, true)
                if (outValue.resourceId != 0) {
                    headerView.setBackgroundResource(outValue.resourceId)
                } else {
                    headerView.setBackgroundColor(outValue.data)
                }

                headerView.elevation = 8f
            }
        }
    }

    override fun isHeader(itemPosition: Int): Boolean {
        return getItemViewType(itemPosition) == TYPE_NODE
    }

    override fun onHeaderClicked(headerPosition: Int) {
        if (headerPosition != -1 && headerPosition < visibleItems.size) {
            val item = visibleItems[headerPosition]
            if (item.item is DtsNode) {
                val node = item.item as DtsNode
                node.isExpanded = !node.isExpanded
                rebuildVisibleList()

                // If we collapsed it, we might be way down the list. Scroll to the header.
                if (!node.isExpanded) {
                    scrollToResult(item)
                }
            }
        }
    }

    // ============================================================================================
    // View Holders
    // ============================================================================================

    inner class NodeViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val indentation: View = itemView.findViewById(R.id.node_indentation)
        private val expandIcon: ImageView = itemView.findViewById(R.id.btn_expand)
        val nameText: TextView = itemView.findViewById(R.id.node_name)

        init {
            itemView.setOnClickListener { toggleExpand() }
            expandIcon.setOnClickListener { toggleExpand() }
        }

        private fun toggleExpand() {
            val pos = bindingAdapterPosition
            if (pos == RecyclerView.NO_POSITION) return

            val item = visibleItems[pos]
            val node = item.item as DtsNode
            node.isExpanded = !node.isExpanded

            rebuildVisibleList()
        }

        fun bind(node: DtsNode, depth: Int, isMatch: Boolean) {
            indentation.layoutParams.width = depth * 40
            indentation.requestLayout()

            if (isMatch && searchQuery.isNotEmpty()) {
                nameText.text = highlightText(node.name, searchQuery)
            } else {
                nameText.text = node.name
            }
            expandIcon.rotation = if (node.isExpanded) 0f else -90f

            if (currentSearchIndex != -1 && searchResults.size > currentSearchIndex && searchResults[currentSearchIndex].item == node) {
                itemView.setBackgroundColor(Color.parseColor("#4Dffeb3b"))
            } else {
                val outValue = TypedValue()
                itemView.context.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
                itemView.setBackgroundResource(outValue.resourceId)
            }
        }
    }

    inner class PropertyViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val indentation: View = itemView.findViewById(R.id.prop_indentation)
        private val keyText: TextView = itemView.findViewById(R.id.prop_key)
        private val valueInput: EditText = itemView.findViewById(R.id.prop_value)
        private var currentWatcher: TextWatcher? = null

        fun bind(prop: DtsProperty, depth: Int, isMatch: Boolean) {
            indentation.layoutParams.width = depth * 40
            indentation.requestLayout()

            if (isMatch && searchQuery.isNotEmpty()) {
                keyText.text = highlightText(prop.name, searchQuery)
            } else {
                keyText.text = prop.name
            }

            if (currentWatcher != null) {
                valueInput.removeTextChangedListener(currentWatcher)
            }

            val displayVal = prop.displayValue
            valueInput.setText(displayVal)

            currentWatcher = object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
                override fun afterTextChanged(s: Editable?) {
                    prop.updateFromDisplayValue(s.toString())
                }
            }
            valueInput.addTextChangedListener(currentWatcher)

            if (currentSearchIndex != -1 && searchResults.size > currentSearchIndex && searchResults[currentSearchIndex].item == prop) {
                itemView.setBackgroundColor(Color.parseColor("#4Dffeb3b"))
            } else {
                val outValue = TypedValue()
                itemView.context.theme.resolveAttribute(android.R.attr.selectableItemBackground, outValue, true)
                itemView.setBackgroundResource(outValue.resourceId)
            }
        }
    }

    private fun highlightText(text: String, query: String?): SpannableString {
        val spannable = SpannableString(text)
        if (query.isNullOrEmpty()) return spannable

        val lowerText = text.lowercase(java.util.Locale.ROOT)
        val lowerQuery = query.lowercase(java.util.Locale.ROOT)

        var start = lowerText.indexOf(lowerQuery)
        while (start != -1) {
            val end = start + lowerQuery.length
            spannable.setSpan(BackgroundColorSpan(Color.parseColor("#ffeb3b")), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            spannable.setSpan(ForegroundColorSpan(Color.BLACK), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            start = lowerText.indexOf(lowerQuery, end)
        }
        return spannable
    }

    class TreeItem(val item: Any, val depth: Int) {
        var isMatch: Boolean = false
    }
}
